### PR TITLE
Add support and stability level definitions

### PR DESCRIPTION
**Pull request by jpkrohling**
Proposal for the definition of the stability and support levels. Once we have this in place, I'll add them to a couple of components to be used as example.

Notes:
* I thought about adding specific requirements to promote a component from experimental to stable, but in the end, I think code owners should be free to determine when a component is ready to be declared stable
* In the support levels, there's a mention that the maintainers are ultimately responsible for all components. This deliberately includes vendor-specific ones, as we might be forced to remove vendor-specific components
* To be proposed in the next step, but wanted to give a heads-up: the individual components could list the code owners, so that users know who to ping in case of problems. So, the support is still provided by the maintainers but specific people (non-maintainers) could be pinged directly.

Fixes #9379

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

### CODE REVIEW

File: README.md
Comment (bogdandrutu) [inline README.md:61]: Do we need a beta level as well, or RC?

File: README.md
Comment (bogdandrutu) [inline README.md:55]: Should we have a way to expose this on the zpages or with a special command on the otelcol binary?

File: README.md
Comment (bogdandrutu) [inline README.md:63]: Is there a level as well here? or just a support "party"?

File: README.md
Comment (mx-psi) [inline README.md:55]: Do we want a 'Deprecated' stability level or is that out of scope?

File: README.md
Comment (mx-psi) [inline README.md:53]: Should we have a template table/some kind of 'copy and fill in the blanks' to add on components' READMEs?

File: README.md
Comment (mx-psi) [inline README.md:55]: What happens if a component is stable for a telemetry signal but not for others?  For example, how does the current status of the OTLP receiver defined [here](https://github.com/open-telemetry/opentelemetry-collector#status) map to these levels? I think it would not make sense to consider it stable unless logs functionality was under a feature flag.

Do we also want to make a distinction between configuration stability and behavior stability (say, do we want to allow expressing 'this component behavior is stable but the way it is configured may change in breaking ways')?

File: README.md
Comment (mx-psi) [inline README.md:61]: Should we explicitly mention that functionality exposed under feature gates is not covered by these stability guarantees?

Comment (jpkrohling) [reply README.md:61]: How would you describe them both? In my view, experimental encompasses both, and component maintainers would be free to move from experimental to stable as they see fit.

File: README.md
Comment (jpkrohling) [reply README.md:55]: Good call. I think @esnible added an option to print out all the components of the collector upon error, perhaps a code similar to that would be good to have here. I'll create another issue to track this.

Comment (jpkrohling) [reply README.md:55]: Created #9518 and #9519.

File: README.md
Comment (jpkrohling) [reply README.md:63]: I think we should start with a stability level and support party. Those supporting a component are the ones defining the stability level, so, they can adjust the expectations as needed.

File: README.md
Comment (jpkrohling) [reply README.md:55]: Great call, we need deprecated there.

File: README.md
Comment (jpkrohling) [reply README.md:53]: I'm hoping one of the next PRs will establish the template. From there, we can determine whether it's easy enough for other components to copy/paste, or whether a template somewhere would be better.

File: README.md
Comment (jpkrohling) [reply README.md:55]: We definitely need a dimension for the signal type, but I don't think it's a good user-experience to break it even more in config vs. behavior stability. From the user's perspective, they are all part of the same contract, IMO.

File: README.md
Comment (jpkrohling) [reply README.md:61]: I'll add a paragraph about it.

File: README.md
Comment (TylerHelmuth) [inline README.md:49]:
```suggestion
This is a repository for OpenTelemetry Collector components that are not suitable for the  [core repository](https://github.com/open-telemetry/opentelemetry-collector) of the collector. Some of these components are part of the core distribution, such as the Jaeger and Prometheus components, but most of the components here are only available as part of the "contrib" distribution. Users of the OpenTelemetry Collector are also encouraged to build their own custom distributions with the [OpenTelemetry Collector Builder](https://github.com/open-telemetry/opentelemetry-collector/tree/main/cmd/builder), using the components they need from the core repository, the contrib repository, and possibly third-party or internal repositories.
```

File: README.md
Comment (jpkrohling) [reply README.md:61]: Can you check it out?

File: README.md
Comment (jpkrohling) [reply README.md:55]: I added a few words about it, would you mind taking a look?

**Comment by jpkrohling** [comment]
I believe I addressed all comments, but there are some discussions pending.

File: README.md
Comment (mx-psi) [inline README.md:57]:
```suggestion
While we intend to provide high-quality components as part of this repository, we acknowledge that not all of them are ready for prime time. As such, each component should list its current stability level for each telemetry signal, according to the following definitions:
```

File: README.md
Comment (mx-psi) [reply README.md:61]: Current wording looks good, I think we should also link to https://github.com/open-telemetry/opentelemetry-collector/blob/main/service/featuregate/README.md#feature-lifecycle to provide more context

File: README.md
Comment (mx-psi) [reply README.md:55]: Looks good and agree that the distinction between configuration and behavior may not be a good user experience. I left a comment below but we can resolve this one!

File: README.md
Comment (mx-psi) [reply README.md]: 👍 added https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9377#issuecomment-1109510746

File: README.md
Comment (pmm-sumo) [inline README.md:59]: Should we define a state which precedes "In development"? E.g. when component is not ready to be used in any way and has only interface available (during the good practices of breaking the large work into chunks). Something like "Do not use", "Bootstrapping" or so?

Comment (jpkrohling) [reply README.md:59]: That's precisely what I had in mind for "In development" -- the component is likely not in a working state, it hasn't been finished, and might not even be tied to the factories yet.

File: README.md
Comment (jpkrohling) [reply README.md:61]: I would read the alpha stage as "Experimental" here, while their beta would map to "Stable". I don't think the beta would be a stage in between, given the definitions from this PR.

Comment (jpkrohling) [reply README.md:61]: Perhaps it would be better to rename "Experimental" to "Beta"?

**Comment by jpkrohling** [comment]
Just to make it explicit how I see the lifecycle of a component to be, which would justify the need for only three stability levels:

- a component is accepted into this repository and development starts. There might be multiple PRs until the component is ready for consumption. This is the "in development" level.
- once the development has been finished, the component owners make the component available for people to start using it and provide feedback. Real world usage starts here, with people reporting bugs and asking for new features (better observability, new capabilities, ...). This is the "experimental" level.
- once the component has been in the wild for a while and component owners are confident it is ready for production usage, the component is marked as "stable".

As a user, I'm not sure I see the need for the second stage to be broken down into other stages: the information I need is only whether the component I need is ready for production or not.

**Comment by TylerHelmuth** [comment]
> As a user, I'm not sure I see the need for the second stage to be broken down into other stages: the information I need is only whether the component I need is ready for production or not.

I tend to agree.  Though if the code owners want to breakdown the Experimental stage into more granular levels within their README somewhere else I think that is ok, as long as it is also marked Experimental in all expected places (top of README, any tables somewhere, etc).

File: README.md
Comment (dmitryax) [reply README.md:61]: Maybe breaking changes to the configuration interface can be used as one of differentiator between Experimental and Beta? Current configuration guarantees for Experimental seem too strict IMO.

E.g. **Beta** is ready for production use but the configuration options might change between releases, although component owners should try to minimize breaking changes.

While **Experimental** configuration options might break often depending on how things evolve.

Comment (jpkrohling) [reply README.md:61]: Alright, I renamed Experimental to Alpha and added a Beta. As a user, I still think I would prefer to have a single "yes/no" on whether I should use it in production, as the message here is now:

- yes, if it's stable
- maybe, if it's beta (your config might break)
- no, if it's alpha

File: README.md
Comment (bogdandrutu) [inline README.md:67]: I think the correct answer is, look for the lifecycle stage? We should not consider any feature behind a gate as experimental.

Also I like that our stages map with https://github.com/open-telemetry/opentelemetry-collector/blob/v0.49.0/service/featuregate/README.md#feature-lifecycle where GA == stable.

File: README.md
Comment (bogdandrutu) [reply README.md:71]: I am asking about the title, showed we call this "Support level" or can we find a different word? My fear is to not confuse the "stability **level**" with "support **level**"

File: README.md
Comment (TylerHelmuth) [reply README.md:59]: What about components that are in the Contrib repository but not part of the contrib distribution (transform processor is an example I believe).  Should it be marked as "In Developement" or is there something even lower?

**Comment by TylerHelmuth** [comment]
> > As a user, I'm not sure I see the need for the second stage to be broken down into other stages: the information I need is only whether the component I need is ready for production or not.
> 
> I tend to agree. Though if the code owners want to breakdown the Experimental stage into more granular levels within their README somewhere else I think that is ok, as long as it is also marked Experimental in all expected places (top of README, any tables somewhere, etc).

Reading through the change from Experimental to Alpha/Beta I actually think I've changed my mind.  Alpha and Beta are well defined and Stable is still the indicator that "yes, this is good to use in production".  As a customer (especially if I am an early adopter) Alpha and Beta will make it easier to decide which "Experimental" components I would feel comfortable using in production.  When it was only "Experimental" I would probably have to dig deeper into the community/code to understand how risky it really is.

File: README.md
Comment (TylerHelmuth) [reply README.md]: Could probably get away with only "Support"

File: README.md
Comment (TylerHelmuth) [inline README.md]: If kept, Experimental should be updated to either Alpha or Beta

File: README.md
Comment (TylerHelmuth) [reply README.md:59]: "it might not be available as part of any distributions yet"  answer was already in there.

File: README.md
Comment (pmm-sumo) [reply README.md]: So if there's e.g. a stable component that has a feature gate, the said feature should be considered "**Alpha**" rather than "**Experimental**"?

Comment (TylerHelmuth) [reply README.md]: I think the whole line was actually removed based on @bogdandrutu's feedback and this comment somehow didn't marked "Outdated"

File: README.md
Comment (pmm-sumo) [inline README.md:49]: > Some of these components are part of the core distribution, such as the Jaeger and Prometheus components

Not sure if I read it right but Jaeger and Prometheus were moved from core to contrib quite a while ago

Comment (TylerHelmuth) [reply README.md:49]: Thats true, but that are still included in the core distribution of the collector: https://github.com/open-telemetry/opentelemetry-collector-releases/blob/main/distributions/otelcol/manifest.yaml

Comment (pmm-sumo) [reply README.md:49]: So what I'm thinking is that perhaps this would be more clear in what's the relationship between core/contrib repositories/releases

```suggestion
This is a repository for OpenTelemetry Collector components that are not suitable for the  [core repository](https://github.com/open-telemetry/opentelemetry-collector) of the collector. 

The official distributions, core and contrib, are available as part of the [opentelemetry-collector-releases](https://github.com/open-telemetry/opentelemetry-collector-releases) repository. Some of the components in this repository are part of the "core" distribution, such as the Jaeger and Prometheus components, but most of the components here are only available as part of the "contrib" distribution. Users of the OpenTelemetry Collector are also encouraged to build their own custom distributions with the [OpenTelemetry Collector Builder](https://github.com/open-telemetry/opentelemetry-collector/tree/main/cmd/builder), using the components they need from the core repository, the contrib repository, and possibly third-party or internal repositories.
```

**Review by mx-psi — APPROVED** [review]

**Review by pmm-sumo — APPROVED** [review]
I left [a comment on core distro wording](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/9412#discussion_r860171010) but overall this LGTM

**Comment by jpkrohling** [comment]
I applies some suggestions but forgot to publish the review comments :-) I have also applied @pmm-sumo's suggestion about the distributions/repositories.

In any case, I think you all have seen the changes already, but I'll give it a few hours for folks to give it a final look.

**Review by bogdandrutu — APPROVED** [review]

File: README.md
Comment (pmm-sumo) [inline README.md:53]:
```suggestion

```
After merging my previous suggestion (should have used a multi-line one) this is no longer necessary

File: README.md
Comment (dmitryax) [inline README.md:63]: Didn't you propose that **Beta** can be used in production while config might break? https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/9412#discussion_r860038480

**Review by Aneurysm9 — APPROVED** [review]

File: README.md
Comment (jpkrohling) [reply README.md:63]: I had only the config changes in mind based on previous comments, but we could say that people can have higher confidence in a beta component to run it with more workloads than they'd run with an alpha component. Let me draft something to that effect.

Comment (jpkrohling) [reply README.md:63]: > A component at this stage is expected to have had exposure to non-critical production workloads already during its **Alpha** phase, making it suitable for broader usage.

How about this?

Comment (TylerHelmuth) [reply README.md]: I think that wording is good.  I like keeping "Stable" as the only place to official say "this is officially GA".

Comment (TylerHelmuth) [reply README.md]: Although now I'm wondering if instead of "**Stable**: the component is ready to be used in production." we should say something like "**Stable**: the component is ready for general availability."  The form implies no stage before "Stable" should ever be used in production, but there are times Beta could be used.  The later implies this component has had a "production" release, thus making it safe to use.

Comment (jpkrohling) [reply README.md]: Makes sense to me

Comment (jpkrohling) [reply README.md]: PR updated.

**Review by jamesmoessis — APPROVED** [review]

**Comment by jpkrohling** [comment]
Looks like we have a good number approvals, so I'll go ahead and merge this.

**Pull request merged**
